Fix #53: Update Tests to probot v9 and fix all test issues

jameesjohn opened the pull request:
<!--
  - Thanks for submitting code to Oppiabot! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #53: Fixes the issue with running tests and upgrade the test to probot v9.
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

jameesjohn commented:
PTAL @ankita240796
Thanks

ankita240796 commented on fixtures/pullRequestPayload.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "event": "pull_request",
+  "name": "pull_request",
```
Why this change?

jameesjohn replied:
This change is made because probot expects to receive a webhook with an object of the structure 
`
{  
name: <event_type>, 
payload: <webhookpayload>
}
`
 https://github.com/oppia/oppiabot/blob/9eaf808aacfe8f8ff97ce46a9b7a0182a3616ae8/spec/oppiabotSpec.js#L110

ankita240796 replied:
Ok, sg.

ankita240796 commented on spec/oppiabotSpec.js:
```diff
@@ -1,22 +1,29 @@
 require('dotenv').config();
-
-const { createRobot } = require('probot');
+const {createProbot} = require('probot');
```
Add spaces after braces: { createProbot }

jameesjohn replied:
Done.

ankita240796 commented on spec/oppiabotSpec.js:
```diff
@@ -102,24 +119,18 @@ describe('Oppiabot\'s', () => {
+          'testuser7777');
       });
 
-    it('should be called for a pull request for the given payload', () => {
-      expect(
-        apiForSheetsModule.checkClaStatus.calls.argsFor(0)[2])
-        .toEqual(true);
-    });
-
```
Why is this removed?

jameesjohn replied:
I'm not sure what was being tested here because `apiForSheetsModule` expects just one argument, and the check is testing for the third argument.

jameesjohn replied:
I'm not sure what we should test here.

ankita240796 replied:
I think the test is related to contextData, can you try to figure out what was being compared with true in the original test.

jameesjohn replied:
@ankita240796 I think it checks that the data gotten is a pull request data
https://github.com/oppia/oppiabot/blob/3aa31ef80b5af0789e34d608ef85ec9a11becd61/index.js#L32
https://github.com/oppia/oppiabot/blob/3aa31ef80b5af0789e34d608ef85ec9a11becd61/lib/apiForSheets.js#L114

ankita240796 replied:
So, we should test the same in the current tests as well.

jameesjohn replied:
I don't think it is necessary to check if it is a pull request data or not since we are only calling the `apiForSheetsModule.checkClaStatus` function when a pull request is opened.
I believe this was done initially because the checkClaStatus was also called when an issue comment was created.
https://github.com/oppia/oppiabot/blob/511f4f06332f73b2efa9a258305d1821d3aaacac/index.js#L15

ankita240796 replied:
Ok, sg.

ankita240796 commented on spec/oppiabotSpec.js:
```diff
@@ -102,24 +119,18 @@ describe('Oppiabot\'s', () => {
     });
 
-    it('should be called with three arguments for the given payload', () => {
+    it('should be called with one argument for the given payload', () => {
```
Can you please explain this change?

jameesjohn replied:
`apiForSheetsModule.checkClaStatus` expects to be called with just one argument `Context`
https://github.com/oppia/oppiabot/blob/04d7f3b8723165bbbdc51074855d5223af75b2cf/index.js#L27

ankita240796 replied:
Ok, sg.

jameesjohn commented:
Hi @ankita240796, I've responded to comments, PTAL. Thanks!

jameesjohn commented:
> Regarding running tests locally, I think we should have clear instructions and a python script... 


@ankita240796, would it be okay if I write the script in node?

ankita240796 commented:
> > Regarding running tests locally, I think we should have clear instructions and a python script...
> 
> @ankita240796, would it be okay if I write the script in node?

Yes, any language is fine as long as we have clear instructions on how to run.

jameesjohn commented:
PTAL @ankita240796
I've created a `setup.js` file which will handle updating the `.env`, running the tests and reverting it back once the tests are done running.

jameesjohn commented:
PTAL @ankita240796. I've responded to all comments.

ankita240796 commented on lib/apiForSheets.js:
```diff
@@ -188,6 +188,8 @@ module.exports.checkClaStatus = function(contextData) {
   // Authorize a client with the loaded credentials, then call the
   // Google Sheets API.
-  module.exports.authorize(
+  const oauthClient = module.exports.authorize(
     JSON.parse(clientSecret), module.exports.checkClaSheet);
+
+  module.exports.checkClaSheet(oauthClient);
 };
```
Can you explain the purpose of this change i.e. why is it required for running the tests?

ankita240796 replied:
Thanks for the explanation!

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable no-console */
+const fs = require('fs');
```
Let us add instructions on how to run this as well as a copyright license here and everywhere else.

Also, can you provide a set of points on what you have tried and what are the issues you are facing with newrelic while running the tests (You can create a doc as well).

jameesjohn replied:
This can be run by running `npm run test`.
I added it to the test script in package.json https://github.com/oppia/oppiabot/blob/6c02cbd51d7bec70a8ced69fa26770c69ebaae16/package.json#L12

ankita240796 replied:
Hmm, so the test is written in a way that considers `.env` is already present in the oppiabot folder. Let us update it so that it creates `.env` using `.env.example` if it is missing.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+let envData = '';
+
+const setWhitelistedAccount = () => {
+  // Load env
```
End in period.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+  });
+  envData = data.toString();
+
+  // Parse and remove comments
```
End in period.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+    }
+  });
+
+  // check for whtelisted_account
```
Capitalize first letter, end in period.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+  );
+  const whitelistIndex = envArray.indexOf(whitelist);
+  if (!whitelist.includes('oppia')) {
+    // Replace the current whitelist with new one
```
End in period.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+
+  // check for whtelisted_account
+  const whitelist = envArray.find((line) =>
+    line.startsWith('WHITELISTED_ACCOUNTS')
```
`WHITELISTED_ACCOUNTS`: Declare this as a constant and use everywhere.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+
+  // set client secret
+  const clientSecretIndex = envArray.findIndex((line) =>
+    line.startsWith('CLIENT_SECRET')
```
Ditto.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,85 @@
+
+const runTest = () => {
+  const jasminePath = path.join(
+    __dirname,
+    '..',
+    'node_modules',
+    '.bin',
+    'jasmine'
+  );
```
Move in single line.

jameesjohn replied:
Keeping them all on one line is longer than 80 lines and eslint complains.

ankita240796 replied:
You can split into two lines.

jameesjohn replied:
Done.

jameesjohn commented:
I've responded to all comments, PTAL @ankita240796. Thanks!

ankita240796 commented:
Hey @jameesjohn, left a few more comments, PTAL. Thanks!

jameesjohn commented:
Done @ankita240796, PTAL. Thanks!

ankita240796 commented:
Hmm ok, so I see that the tests fail since newrelic config is not found. Is this the same issue you are facing?

jameesjohn commented:
No
Mine passes without the new relic config.
Can you send a screenshot?

ankita240796 commented:
Here is the log:

```
Updating .env
Error: Command failed: "/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/.bin/jasmine"
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

    at ChildProcess.exithandler (child_process.js:303:12)
    at ChildProcess.emit (events.js:311:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Socket.<anonymous> (internal/child_process.js:443:11)
    at Socket.emit (events.js:311:20)
    at Pipe.<anonymous> (net.js:668:12) {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/.bin/jasmine"'
}
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

Started
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.166Z","msg":"Using New Relic for Node.js. Agent version: 4.13.1; Node version: v12.16.1."}
{"v":0,"level":40,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.167Z","msg":"New Relic for Node.js 4.13.1 has not been tested on Node.js v12.16.1. Please update the agent or downgrade your version of Node.js"}
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.167Z","msg":"Module not enabled in configuration; not starting."}
FFFFFFFFFFFFF

Failures:
1) Oppiabot's apiForSheets should be called for the given payload
  Message:
    TypeError: createProbot is not a function
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:93:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
  Message:
    TypeError: Cannot read property 'receive' of undefined
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:110:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
  Message:
    Expected spy checkClaStatus to have been called.
  Stack:
    Error: Expected spy checkClaStatus to have been called.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:115:49)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
```

jameesjohn commented:
Uhmm...
Which version of node.js are you using?

When I run the test, it gives the new relic error message, but it continues to run the tests.

jameesjohn commented:
```
Updating .env
Unable to find New Relic module configuration. A default
configuration file can be copied from C:\Users\PT WORLD\opensource\oppiabot\node_modules\newrelic\lib\config\default.js
and put at C:\Users\PT WORLD\opensource\oppiabot\newrelic.js or
C:\Users\PT WORLD\opensource\oppiabot\node_modules\jasmine\bin\newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from C:\Users\PT WORLD\opensource\oppiabot\node_modules\newrelic\lib\config\default.js
and put at C:\Users\PT WORLD\opensource\oppiabot\newrelic.js or
C:\Users\PT WORLD\opensource\oppiabot\node_modules\jasmine\bin\newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

Started
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"JamesSpectre","pid":14208,"time":"2020-05-04T16:28:25.871Z","msg":"Using New Relic for Node.js. Agent version: 4.13.1; Node version: v10.16.3."}
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"JamesSpectre","pid":14208,"time":"2020-05-04T16:28:25.873Z","msg":"Module not enabled in configuration; not starting."}
16:28:27.587Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.602Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.636Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.648Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.707Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.723Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.754Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.768Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.782Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.812Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.828Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.858Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.882Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.


13 specs, 0 failures
Finished in 0.596 seconds

Reverting .env
```

ankita240796 commented:
Ok, so let us do two things:

1. Let us set this var `NEW_RELIC_NO_CONFIG_FILE` to true in env based on the message here: 
```
If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
```

2. Let us have a script which checks and installs the required versions of all the deps before running the tests (as we have in oppia repo).

ankita240796 commented on lib/apiForSheets.js:
```diff
@@ -188,6 +188,7 @@ module.exports.checkClaStatus = function(contextData) {
   // Google Sheets API.
-  module.exports.authorize(
-    JSON.parse(clientSecret), module.exports.checkClaSheet);
+  const oauthClient = module.exports.authorize(JSON.parse(clientSecret));
+
+  module.exports.checkClaSheet(oauthClient);
```
Let us add an explanation here as well.

jameesjohn replied:
Done

ankita240796 commented on package.json:
```diff
@@ -9,13 +9,13 @@
     "lint": "standard",
     "start": "probot run ./index.js",
-    "test": "./node_modules/.bin/jasmine"
+    "test": "node ./spec/setup.js"
```
Can we do this change for package.json in a separate PR?

jameesjohn commented:
> Hi @jameesjohn, I am still not able to run the tests

Hi @ankita240796, did you run `npm install`? Since newrelic got updated, I think you need to reinstall it.

> why do we specifically need to remove package-lock.json from .gitignore?

We need package-lock.json committed so that we can ensure that the same version of dependencies are installed by all the contributors.

ankita240796 commented:
Hi @jameesjohn, yes it works fine now, I missed updating by running `npm install`. Taking a final pass now. Thanks!

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,126 @@
+  const newRelicConfigIndex = envArray.findIndex((line) =>
+    line.startsWith(NEW_RELIC)
+  );
+  const newRelicConfig = 'NEW_RELIC_NO_CONFIG_FILE=true';
```
Interpolate the constant here.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,126 @@
+  const newRelicAppIndex = envArray.findIndex((line) =>
+    line.startsWith(NEW_RELIC_APP)
+  );
+  const newRelicConfigApp = 'NEW_RELIC_APP_NAME=oppiabot';
```
Ditto.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,126 @@
+};
+
+const runTest = () => {
+  const jasminePath = path.join(__dirname, '..', 'node_modules',
```
Break after opening bracket.

jameesjohn replied:
Done.

ankita240796 commented on spec/setup.js:
```diff
@@ -0,0 +1,126 @@
+  if (fs.existsSync(envPath)) {
+    data = fs.readFileSync(envPath);
+  } else {
+    data = fs.readFileSync(envExamplePath);
```
Have a check here: if example file is missing as well, throw an error asking the user to update their branch.

jameesjohn replied:
Done.

jameesjohn commented:
Hi @ankita240796, I've addressed all the review comments, PTAL. Thanks!

> We should set up coverage check as well either in this PR or a separate one

Yes, I agree. I think it should be in a separate PR.

ankita240796 commented on spec/setup.js:
```diff
@@ -35,8 +35,11 @@ const setWhitelistedAccount = () => {
     data = fs.readFileSync(envExamplePath);
+  } else {
+    throw new Error(
+      '.env.example could not be found. Please update your branch with master');
```
End with Period.

jameesjohn replied:
Done.
